frmts/sentinel2/sentinel2dataset.cpp

```diff
@@ -438,6 +440,33 @@ int SENTINEL2Dataset::Identify( GDALOpenInfo *poOpenInfo )
         strstr(pszHeader, "User_Product_Level-2A" ) != nullptr )
         return TRUE;
 
+    /* Last effort – look inside of a zip file for metadata file with
+     * matching name. */
+    if ( EQUAL(CPLGetExtension(pszJustFilename), "zip") )
```

Comment:
I'm not super enthusiastic about having to list the content of arbitrary zip files in a Identify() method. Wouldn't it be an option for your use case that the user prefixes the zip filename with a "SENTINEL2:" prefix ? (you'd need to make change accordingly to detect that prefix and strip it out in Open())

Reply:
I can see your point but that would not work with drag&drop approach taken by QGIS.

Reply:
I'm thinking to an alternative that should work and avoid doing VSIOpenDir() for unrelated .zip
If you open a .zip file with Sentinel2 content in a binary editor, you can see in the first bytes of the file, the name of the first file inside the zip. So it should be possible to look into the buffer poOpenInfo->pabyHeader of size = poOpenInfo->nHeaderBytes if it contains one of the substrings tested by SENTINEL2isMetadataFile(). If so, then use the below approach to confirm. Otherwise just give up.

Reply:
You are right. I implemented a new version that uses only header for identification and unzipped version only when opening.